readthedocs/search/parsers.py

```diff
@@ -145,16 +145,54 @@ def _parse_sections(self, title, body):
             tags = body.css(f'h{head_level}')
             for tag in tags:
                 try:
-                    title, id = self._parse_section_title(tag)
+                    title, _id = self._parse_section_title(tag)
                     next_tag = self._get_header_container(tag).next
                     content, _ = self._parse_section_content(next_tag, depth=2)
                     yield {
-                        'id': id,
-                        'title': title,
-                        'content': content,
+                        "id": _id,
+                        "title": title,
+                        "content": content,
                     }
                 except Exception as e:
-                    log.info('Unable to index section.', section=str(e))
+                    log.info("Unable to index section.", section=str(e))
+
+        # All terms in dls are treated as sections.
+        dls = body.css("dl")
+        for dl in dls:
+            # Select all dts with id defined
+            dts = dl.css('dt[id]:not([id=""])')
+
+            # https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dt
+            # multiple <dt> elements in a row indicate several terms that are
+            # all defined by the immediate next <dd> element.
+            for dt in dts:
+                try:
+                    title, _id = self._parse_dt(dt)
+                    # https://developer.mozilla.org/en-US/docs/Web/CSS/General_sibling_combinator
+                    dd = dt.css_first("dt ~ dd")
+                    # We only index a dt with an id attribute and an accompanying dd
+                    if not dd or not _id:
+                        continue
+
+                    # The content of the <dt> section is the content of the accompanying <dd>
+                    content = dd.text()
```

Comment:
This should still run some of the logic from `_parse_section_content` (but checking for the next definition list, since we can have nested definitions https://requests.readthedocs.io/en/latest/api/#requests.Response), since anything can be inside a dd, in case that there is a numbered code block we want to run the logic from `_parse_code_section`.

Reply:
We shouldn't need any recursive logic here since it's handled with `dls = body.css("dl")` ? I looked at `_parse_section_content` and it calls `is_section` which is not exactly up2date and thinks that only headers are sections --- so it won't find numbered code blocks. I think we should detect "code block sections" independently from how `dl` parsing works.

It's interesting that we are creating a flat structure from a nested structure. And that will almost certainly lead to decisions that we have to make. I think the current implementation is a nice and simple addition. But I would want to slightly rewrite things so it's easier to quickly grasp how the call pattern inside the parser works.

Reply:
> We shouldn't need any recursive logic here since it's handled with dls = body.css("dl") ? 

The problem is that it will duplicate the content of nested lists, giving users duplicate results, so I'm suggesting doing something similar to what we did with nested sections.

> I looked at _parse_section_content and it calls is_section which is not exactly up2date 

Yes, that's why I'm suggesting using something similar, to detect definition lists instead of sections for this case.

Reply:
> giving users duplicate results

but isn't that desirable, given that a nested text node is also inside the outer nesting section? I don't see that it's such a big problem.

I find it easy to come up with an example where it seems desirable?

```html
<!-- Look at how generic and uninteresting this section's title is... if we skip the <dl>, there's nothing to index -->
<h2 id="...">Reference</h2>

<dl>
<dt id="...">Very unique and great search term</dt>
<dd>
   <dl>
     <dt id="...">Definition 1></dt>
     <dd>Oh here is some really important text that won't be indexed by neither of the two nesting sections?</dd>
   </dl>
</dd>
</dl>
```

Reply:
The result of B is what we want, we transform nested sections/lists into a flat sequence of lists/sections.

Reply:
Check, but then I'm kind of fine with the duplicate outputs.

When you removed `<dl>` lists from the previous sections, was it because that typical Sphinx domains were too much content to have in those sections?

Reply:
> but isn't that desirable, given that a nested text node is also inside the outer nesting section? I don't see that it's such a big problem.

In your example, that text will be returned 3 times when searching, 3 results for the same block of text.

Reply:
@stsewd something about this feels like it would be a classic search index problem :)

But can you also see the point that when the contents are removed from the containing/nesting element, then the indexing of that element is "worsened" or "less representative" ?

Reply:
Another way of solving this without having to use recursion to detect nested sections/lists would be to start indexing from the child nodes up to the parent nodes and delete the children after parsing, but I haven't explored this solution.

Reply:
> But can you also see the point that when the contents are removed from the containing/nesting element, then the indexing of that element is "worsened" or "less representative" ?

It can trim the context, yeah, but I think that's okay compared to have duplicate results of the same section/text.

Reply:
> Another way of solving this without having to use recursion to detect nested sections/lists would be to start indexing from the child nodes up to the parent nodes and delete the children after parsing, but I haven't explored this solution.

Opened https://github.com/readthedocs/readthedocs.org/issues/10184

Reply:
Thanks @stsewd - it is really helpful that you broke it out into another issue :+1: 

We should definitely figure this out, but I feel like it's a more general design decision about how indexing, searching and presenting results should work and that it might affect more elements which we want to index.

I'm also not saying that I'm an expert on this, I barely know how the search works.. I just know how to generate the JSON for the index now :)